src/lib/es5.d.ts

```diff
@@ -1139,6 +1144,11 @@ interface Array<T> {
       * @param items Additional items to add to the end of array1.
       */
     concat(...items: ConcatArray<T>[]): T[];
+    /**
+      * Combines two or more arrays.
+      * @param items Additional items to add to the end of array1.
+      */
+    concat<U>(...items: (ConcatArray<U> | ConcatArray<T>)[]): (T | U)[];

```

Comment:
first you do not need the first overload with this change. this one covers it already. second, we intentionally did not add this overload to avoid heterogeneous array creation. we think most ppl think of concat as maintaining the type of the correct array and adding elements of a different type is expected to be an error.

Reply:
Regarding the unnecessary overload: indeed, I'll remove it. Regarding the heterogeneous array creation: the problem with *not* adding this overload is that it can lead to unintended heterogeneous array creation, as shown in the original issue. I don't agree that "most people think of concat as maintaining the type of the array", since it creates a **new array**. People with this kind of "false expectation" will likely get an error when they try to use the created array, so I don't see it as a problem...